### PR TITLE
clean up encoding/decoding traits

### PR DESCRIPTION
**Pull request by apoelstra**
Builds on #264

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
ACK if you make it fucking compile.

File: src/internal_macros.rs [hunk 1]
```diff
@@ -18,18 +18,21 @@
...
+                Ok(len)
             }
         }
 
-        impl<D: ::consensus::encode::Decoder> ::consensus::encode::Decodable<D> for $thing {
+        impl<D: ::consensus::ReadExt> ::consensus::Decodable<D> for $thing {
             #[inline]
             fn consensus_decode(d: &mut D) -> Result<$thing, ::consensus::encode::Error> {
-                use consensus::encode::Decodable;
+                use consensus::Decodable;
```
Comment (stevenroose) [inline src/internal_macros.rs]: Given that all type references in the macros have the `::` prefix, this one should have it too I guess.

File: src/network/constants.rs [hunk 2]
```diff
@@ -102,44 +99,9 @@ impl Network {
...
-}
-
 #[cfg(test)]
 mod tests {
   use super::Network;
-  use consensus::encode::{deserialize, serialize};
-
-  #[test]
-  fn serialize_test() {
-    assert_eq!(serialize(&Network::Bitcoin), vec![0xf9, 0xbe, 0xb4, 0xd9]);
```
Comment (stevenroose) [inline src/network/constants.rs]: Couldn't you have kept the tests, but just added `.magic()`?

**Comment by apoelstra** [comment]
Addressed Steven's nits.

@stevenroose the reason for renaming `Encoder` and `Decoder` is that Matt says he was misled by the old names; for returning the length from `encodable` is so that you can determine the length of hashed data without re-encoding it into some special-purpose counting `Write`; for returning `usize` from size-returning functions is so that you don't need a gazillion casts everywhere to use them.

**Comment by stevenroose** [comment]
And for privatizing Encodable? :)

One argument against it would be someone building Bitcoin extension stuff
that is not yet mature/accepted/whatever so doesn't fit in rust-bitcoin,
but could be an optional extension. Currently you're kind of saying that
"all consensus encodable stuff must live in this library".

**Comment by apoelstra** [comment]
Yes, we are saying that. The argument (from @TheBlueMatt) is that we're not intending to create or support a general serialization library, and "Bitcoin extension stuff" is not Bitcoin.

**Comment by stevenroose** [comment]
I doubt it's that simple. I can envision things like Utreexo that core
doesn't support yet, but someone wants to build a POC node or
special-purpose node that uses it. I guess the point I'm trying to argue is
that it should be enough to document a string discouragement from using the
consensus trait in custom stuff instead of having to make it impossible. I
know one could in theory have `impl <E: bitcoin::consensus::Encodable>
UtreexoExtensionEncodable for E` that trivially forwards, but it's annoying.

I'm not saying that rust-elements should still be using the
bitcoin::consensus trait, necessarily. I'm just saying the added benefit of
prohibiting users of doing something apart from telling them not to do
things is small. Especially when it's a niche trait that "regular"/dumber
users won't even notice.

**Comment by TheBlueMatt** [comment]
You *still* shouldn't be using the Rust encoding traits for that. If you build some demo app, you can trivially build a X : std::io::Write and then call the rust-bitcoin traits to write them when you have a rust-bitcoin object and write the other data manually (which you'd have to do anyway).

**Comment by apoelstra** [comment]
I don't really see how you can "misuse" serialization traits, but I agree they're fairly special-purpose and I'd like to retain the ability to change them (in particular they probably support some unnecessary things, like serializing `i16`) without causing tons of downstream pain.

**Comment by apoelstra** [comment]
In person Matt points out that having an impl for byte-slices (or any slices, really) interacts badly with Rust's implicit coercions, which can convert fixed-width types to variable-width types resulting in surprising serializations.

Given that serde has the same issue (see https://github.com/rust-bitcoin/rust-secp256k1/issues/73 for example) I'm not very convinced by this. Removing the impl for `&[T]`, which happened in #264, solved this.

**Comment by TheBlueMatt** [comment]
Well its also an example of why this framework isnt really designed to be a generic, easy-to-use API (ie the only type of thing we should export) - its easy to be confused as to what will happen. If we want to support it as such we can put in some work to document it properly, make sure its consistent for such issues, then fine, but until then I think its pretty clearly not something that meets our export criteria.

**Comment by TheBlueMatt** [comment]
And if we do want a supported, public serialization framework, it should certainly be its own crate.

**Comment by apoelstra** [comment]
rebased

**Comment by apoelstra** [comment]
cc @stevenroose what needs to be done to make this acceptable to you? I'd like to get this in before releasing a new major version.

@TheBlueMatt if I get rid of the sealing commit is that OK with you?

**Comment by dongcarl** [comment]
What are the benefits and drawbacks of sealing?

**Comment by TheBlueMatt** [comment]
Sealing is an explicit statement (and compile-time checking) that this is *not* an interface which is designed to be implemented by clients, in part because of its ideosynchronicities and in part because it's an indication of Bitcoin data. Clients can easily use the same serializer with the io::Write bindings and their own code to write their types.

**Comment by stevenroose** [comment]
It basically requires all Bitcoin protocol structures to be inside this library and seems like it gives us a veto for protocol stuff to be available to Rust developers. Of course one can work around by implementing a new trait and auto trait it with BitcoinEncodable, but that would break compatibility with other crates potentially.

**Comment by stevenroose** [comment]
I think the "explicit statement that this is *not* an interface which is designed to be implemented by clients" can just also be done in documentation

**Comment by apoelstra** [comment]
I'm indifferent but Matt seems to feel strongly about it (even though the only idiosyncrasy he cites, that `[u8; N]` will silently coerce to `&[u8]`, exists also in serde and has much more subtle/surprising consequences there).

But this PR is blocking a new major release of the library so I'd like to get it in.

**Comment by TheBlueMatt** [comment]
I'm not sure what the use-case is for someone using our serialization framework to serialize their own types. We certainly don't guarantee everything will continue to serialize as-is forever, as we may have to implement different serializers for new Bitcoin types (which would have even more serious implications for downstream users). The only legitimate use-case I could see is for testing new protocol extensions, but if you are going to use them in production (and rust-bitcoin decides, for whatever reason, to not support them), then you almost certainly don't want to be relying on the rust-bitcoin serialization framework.

**Comment by TheBlueMatt** [comment]
As for supporting testing, I don't see why that's not probably better off with a fork of rust-bitcoin? Ultimately either it should be targeting upstream (and, thus, a fork), or it shouldn't be, in which case it should certainly not be using the rust-bitcoin serialization framework.

**Comment by stevenroose** [comment]
Let's say one builds a `bitcoin-p2p` crate that has an interface to send any `ConsensusEncodable` over the wire and let's say some client already implemented some new special block filter messages that rust-bitcoin doesn't have yet (*ugh* murmur *ugh*). He won't be able to send them over the bitcoin-p2p crate's wire, because he can't implement `ConsensusEncodable` for them.

You can say bitcoin-p2p should have a `Wireable` trait and auto-implement it for `ConsensusEncodable`. That'd be a workaround. But I don't really see how that client's special `CompactBlockFilter` type is different from any rust-bitcoin type that is allowed to implement `ConsensusEncodable`.

(Btw, the name is a bit confusing as well. `Message` and `CommandString` etc implement `ConsensusEncodable`, but I'm not sure they are actually used in consensus, they are just part of the wire protocol.)

**Comment by stevenroose** [comment]
@apoelstra If you are anxious to release a version, sure, merge it :) It's not breaking any legit existing usages AFAIK.

**Comment by apoelstra** [comment]
I can't merge it until somebody ACKs it :)

But I'm not in too big of a rush - we also need to get a new version of bitcoin_hashes in, and I'd like to get the merkleblock stuff in, and the BIP158 stuff (though both of those can wait for point releases I believe).

**Comment by TheBlueMatt** [comment]
Things in rust-bitcoin *are* special, because they get updated with the rest of the serialization framework stuff. If we change something in the serialization framework for some new type we're adding, it gets review to make sure it doesn't break anything else, that isn't true for some downstream type. Adding a wrapper is relatively little code and it ensures you don't get broken by upstream changes. I'd argue strongly that using the rust-bitcoin serialization framework is *not* a supported use of the API, but its also *the* obvious way to use it today, which seems very, very, very broken.

**Comment by TheBlueMatt** [comment]
Is it easy to use a fork in Cargo.toml as a dep? ie if you want to add a new message, fork rust-bitcoin, and then set your dependency to apply that as a patch. (if you do so, does that apply to everything else in the dependency tree that also relies on rust-bitcoin, or do you get inconsistent-type-version errors?)

**Comment by dongcarl** [comment]
I agree with the sentiment that our serialization traits should not be used outside the crate, but I don't think we need to programmatically disallow it because:

1. We can put a gigantic warning in the traits' documentation. I don't presume that we have enumerated all the ways that these traits can be useful, so those who _really_ want to footgun can do so at their own (relatively minor) peril.
2. It adds a lot of `impl ::private::Sealed for FooBar { }` lines that makes our code less readable as a whole.

**Comment by TheBlueMatt** [comment]
Hmm, I don't really buy that just putting it in docs is sufficient, but we could add a deprecation warning so that they always see that this is unsupported.

**Comment by dongcarl** [comment]
I'm 100% for deprecated warning.

1. Will get the message across to the user with higher probability than documentation
2. Addresses item no. 2 of https://github.com/rust-bitcoin/rust-bitcoin/pull/265#issuecomment-510118924

**Comment by stevenroose** [comment]
Using a deprecation warning seems reasonable to warn developers implementing the trait.

I'm not 100% sure what the effects of a deprecation warning are. Do you only get it when you implement the trait? Or also if you define a method generically over the trait? Because doing that would not be discouraged in the original `Sealed` implementation, or was it?

**Comment by apoelstra** [comment]
I'll investigate

1. Whether we get deprecation warnings ourselves by using the code in rust-bitcoin;
2. Whether users get deprecation warnings when implementing the traits (which they should not do)
3. Whether users get deprecation warnings when using the trait methods (which they should be able to do)

**Comment by apoelstra** [comment]
Rebased & added two commits to clean up the `Encodable` and `Decodable` traits further.

Will play with deprecation now..

**Comment by apoelstra** [comment]
So, if we deprecate `Encodable` or `Decodable` then we get a gazillion warnings when compiling the library. We can override them at every impl site of course, but that's even more invasive than having `impl ::private::Sealed` at every impl site.

It's also hard to say what the deprecation message should be, given that I'm not aware of any danger associated with implementing either of these traits.

My vote is to neither seal nor deprecate.

**Comment by apoelstra** [comment]
Moved sealing commit to the end to make it easier to drop

**Comment by stevenroose** [comment]
Yeah I was already thinking that deprecating was gonna be ugly. Well if it has to be, just seal them and get it merged. But my vote would still be neither seal nor deprecate.

**Comment by dongcarl** [comment]
I'm okay with not sealing. If we find a lot of people abusing the API and asking us for help, we can reevaluate. I would still like documentation of some sort though.

**Comment by apoelstra** [comment]
What further documentation is needed on any of these traits?

**Comment by dongcarl** [comment]
Let's drop and merge. TBH I don't have strong opinions.

**Comment by apoelstra** [comment]
Dropped.

**Comment by stevenroose** [comment]
ACK 3b9a94a17818d1949c48324af12e00ac6d9d7533

**Comment by dr-orlovsky** [comment]
> I'm not sure what the use-case is for someone using our serialization framework to serialize their own types. 

@TheBlueMatt , let me give you one example. We are working on RGB and Spectrum – assets over Bitcoin and LN: https://github.com/rgb-org/spec and https://github.com/rgb-org/rgb-rust

There we have (a) off-chain proofs and (b) P2P node communications. These proofs contain some transaction data (like public keys, txids etc), which are easy to serialize with bitcoin encoding – but we need to serialize other staff as well. Bitcoin encoding is good for creating really densely-packed data, and that is what we need for off-chain stored histories. Moreover, to communicate across network nodes will serialize PSBTs into the network, so bitcoin encoding will be used there as well.

I understand that the case (a) can't be named "Bitcoin encoding": it only "inherits" some particular encoders from it, and can easily use custom encodings for other types, and be named "RGB encoding" or whatever. However, from the developer point of view, it will be much easier if there a single trait for doing the encoding for your custom structures, and this is the same as bitcoin encoder – you can see some sample code in https://github.com/rgb-org/rgb-rust/blob/master/src/proof.rs#L271

**Comment by tamasblummer** [comment]
@dr-orlovsky you can still use bitcoin::consensus::serialize

utACK

File: src/consensus/encode.rs [hunk 3]
```diff
@@ -393,14 +388,22 @@ pub struct CheckedData(pub Vec<u8>);
...
+            fn consensus_decode<D: io::Read>(mut d: D) -> Result<Self, Error> {
+                ReadExt::$meth_dec(&mut d).map($ty::from_le)
+            }
         }
 
-        impl<S: Encoder> Encodable<S> for $ty {
+        impl Encodable for $ty {
             #[inline]
-            fn consensus_encode(&self, s: &mut S) -> Result<(), self::Error> { s.$meth_enc(self.to_le()) }
+            fn consensus_encode<S: WriteExt>(
```
Comment (TheBlueMatt) [inline src/consensus/encode.rs]: Wait, so now we have WriteExt, but only for a few types, and most just take io::Write? I mean that mostly prevents people from making the mistake I made of trying to implement ReadExt, but not fully. Can we just drop WriteExt (maybe in a future PR)?

Comment (apoelstra) [reply src/consensus/encode.rs]: Everything that implements `io::Write` implements `WriteExt`. No, we cannot "just drop WriteExt"; this trait is needed to encoded Bitcoin types.

**Pull request merged**